### PR TITLE
fix sd loading, don't update affine during eval

### PR DESCRIPTION
**Pull request by theAdamColton**
This is related to https://github.com/pytorch/pytorch/issues/8104

previously, the affine params were not loadable from state dict, because they are initialized in the constructor as None

```python
import torch
from vector_quantize_pytorch import VectorQuantize

vq = VectorQuantize(512, 32, affine_param=True, learnable_codebook=True, ema_update=False)

vq(torch.randn(1,512))

assert vq._codebook.codebook_mean is not None

torch.save(vq.state_dict(), "sd.pt")

vq_loaded = VectorQuantize(512, 32, affine_param=True, learnable_codebook=True, ema_update=False)

# If not using strict mode, you'd get:
#  Unexpected key(s) in state_dict: "_codebook.batch_mean", "_codebook.batch_variance", .....
vq_loaded.load_state_dict(torch.load("sd.pt"), strict=False) 
```

To get around this I initialized the buffers with empty tensors, and added ..._needs_init buffers that determine whether the affine buffers should be updated with ema, or just initialized.

### CODE REVIEW

**Comment by lucidrains** [comment]
💯 thank you!

**Comment by lucidrains** [comment]
@theAdamColton is the affine thing working out for you in preventing codebook collapse?

**Comment by theAdamColton** [comment]
I haven't observed that so far, the orange line is the perplexity with a model using EMA update, the blue line is with affine and learnable codebook 
image

but it would probably be better to compare learnable codebook and affine v.s. learnable codebook and no affine

One thing that I will say is that learnable codebook + affine seems to help a lot with training loss stability compared to EMA.

**Comment by lucidrains** [comment]
isn't lower perplexity better? if you noticed increased stability, that would line up with the paper's claims!

**Comment by theAdamColton** [comment]
higher perplexity is better because it means you have a more of a uniform assignment over the codes, rather than in codebook collapse where only some codes are used

Yeah, I haven't tested enough to say that the affine/learnable codebook loss converges faster than does ema, but the EMA loss curve is usually way more unstable across different learning rates

**Comment by lucidrains** [comment]
@theAdamColton oh i see! i thought it was your end task perplexity

that's cool! you should try the stochastic sampling and see how that fares on your setup 😃

**Comment by lucidrains** [comment]
> higher perplexity is better because it means you have a more of a uniform assignment over the codes, rather than in codebook collapse where only some codes are used
> 
> Yeah, I haven't tested enough to say that the affine/learnable codebook loss converges faster than does ema, but the EMA loss curve is usually way more unstable across different learning rates

nice! i'll take that as +1 anecdata then for stability!

**Comment by gdoras** [comment]
Hello, I was wondering: why do we want to update the batch statistics also during eval ? Does it make any difference ?
